File: +ndi/+cloud/+sync/+internal/downloadNdiDocuments.m
function downloadedNdiDocuments = downloadNdiDocuments(cloudDatasetId, cloudDocumentIds, ndiDataset, syncOptions)
% DOWNLOADNDIDOCUMENTS Downloads a collection of NDI documents and their files.
%
%   ndi.cloud.sync.internal.DOWNLOADEDNDIDOCUMENTS = ndi.cloud.sync.internal.DOWNLOADNDIDOCUMENTS(...
%       CLOUDDATASETID, CLOUDDOCUMENTIDS, NDIDATASET, SYNCOPTIONS)
%
%   This function downloads document metadata from the cloud, and if
%   syncOptions.SyncFiles is true, it also downloads the associated data files
%   to a local staging location and updates document file information.
%   Finally, it adds the documents to the local NDI dataset if provided
%
%   Inputs:
%       cloudDatasetId (1,1) string - The ID of the dataset on the cloud.
%       cloudDocumentIds (cellstr or string array) - A list of
%           cloud-specific document IDs to download.
%       ndiDataset (1,1) ndi.dataset - Optional: The local NDI dataset object.
%       syncOptions (1,1) ndi.cloud.sync.SyncOptions - Synchronization options.
%
%   Outputs:
%       downloadedNdiDocuments (cell): A cell array of the ndi.document objects
%           that were downloaded and added to the dataset.
%
%   See also: ndi.cloud.download.download_document_collection,
%             ndi.cloud.sync.internal.getFileUidsFromDocuments,
%             ndi.cloud.download.download_dataset_files,
%             ndi.cloud.sync.internal.updateFileInfoForRemoteFiles,
%             ndi.cloud.sync.internal.updateFileInfoForLocalFiles

    arguments
        cloudDatasetId (1,1) string
        cloudDocumentIds (1,:) string % string array of document IDs. Cell arrays will be converted.
        ndiDataset ndi.dataset = ndi.dataset.empty
        syncOptions (1,1) ndi.cloud.sync.SyncOptions = ndi.cloud.sync.SyncOptions()
    end
    
    downloadedNdiDocuments = {}; % Initialize to empty cell

    if isempty(cloudDocumentIds)
        if syncOptions.Verbose
            fprintf('No document IDs provided to download.\n');
        end
        return;
    end

    if syncOptions.Verbose
        if cloudDocumentIds == ""
            fprintf('Attempting to download all documents...\n');
        else
            fprintf('Attempting to download %d documents...\n', numel(cloudDocumentIds));
        end
    end

    % 1. Download documents
    % This function should return a cell array of ndi.document objects
    newNdiDocuments = ndi.cloud.download.download_document_collection(cloudDatasetId, cloudDocumentIds);

    if isempty(newNdiDocuments)
        warning('No documents were retrieved from the cloud for the given IDs.\n');
        return;
    end
    if syncOptions.Verbose
        fprintf('Successfully retrieved metadata for %d documents.\n', numel(newNdiDocuments));
    end

    % 2. Handle associated data files
    if syncOptions.SyncFiles
        if syncOptions.Verbose
            fprintf('SyncFiles is true. Processing associated data files...\n');
        end
        if isempty(ndiDataset)
            rootFilesFolder = tempdir();
        else
            rootFilesFolder = ndiDataset.path;
        end

        % Todo: Ensure proper cleanup if anything goes wrong before files
        % are ingested to database.
        filesTargetFolder = fullfile(rootFilesFolder, ndi.cloud.sync.internal.Constants.FileSyncLocation);

        fileUidsToDownload = ndi.cloud.sync.internal.getFileUidsFromDocuments(newNdiDocuments);

        if ~isempty(fileUidsToDownload)
            if syncOptions.Verbose
                fprintf('Found %d unique file UIDs to download for these documents.\n', numel(fileUidsToDownload));
                fprintf('Ensuring download directory exists: %s\n', filesTargetFolder);
            end
            if ~isfolder(filesTargetFolder)
                mkdir(filesTargetFolder);
            end
            
            % This function should download files to the filesTargetFolder
            ndi.cloud.download.download_dataset_files(...
                cloudDatasetId, ...
                filesTargetFolder, ...
                fileUidsToDownload, ...
                "Verbose", syncOptions.Verbose);
            if syncOptions.Verbose
                fprintf('Completed downloading data files.\n');
            end
            
            % Update document file info to point to local files
            if syncOptions.Verbose
                fprintf('Updating document file info to point to local files.\n');
            end
        else
            if syncOptions.Verbose
                fprintf('No associated files found for these documents, or files already local.\n');
                fprintf('Updating document file info (SyncMode.Local, but no new files to point to).\n');
            end
        end
        documentUpdateFcn = @(doc) ...
            ndi.cloud.sync.internal.updateFileInfoForLocalFiles(doc, filesTargetFolder);
    else
        if syncOptions.Verbose
            fprintf('"SyncFiles" option is false. Updating document file info to reflect remote files.\n');
        end
        documentUpdateFcn = @(doc) ...
            ndi.cloud.sync.internal.updateFileInfoForRemoteFiles(doc, cloudDatasetId);
    end

    % 3. Update file info for documents based on local / remote location
    newNdiDocuments = ndi.docs.docfun(documentUpdateFcn, newNdiDocuments);

    % 4. Add documents to the local dataset
    if ~isempty(ndiDataset)
        if syncOptions.Verbose
            fprintf('Adding %d processed documents to the local dataset...\n', ...
                numel(newNdiDocuments));
        end
        ndiDataset.database_add(newNdiDocuments);
        if syncOptions.Verbose
            fprintf('Documents added to the dataset.\n');
        end
    end

    if nargout > 0
        downloadedNdiDocuments = newNdiDocuments;
    end
end

File: ndi-python/ndi/cloud/sync/internal/download_ndi_documents.py
"""Download a collection of NDI documents and their files.

Ported from: ndi.cloud.sync.internal.downloadNdiDocuments (MATLAB)
"""

import os
from typing import List, Optional, TYPE_CHECKING

from ndi.cloud.sync.internal.constants import Constants
from ndi.cloud.sync.internal.get_file_uids_from_documents import get_file_uids_from_documents
from ndi.cloud.sync.internal.update_file_info_for_local_files import update_file_info_for_local_files
from ndi.cloud.sync.internal.update_file_info_for_remote_files import update_file_info_for_remote_files

if TYPE_CHECKING:
    from ndi.dataset import Dataset
    from ndi.document import Document
    from ndi.cloud.sync.sync_options import SyncOptions


def download_ndi_documents(
    cloud_dataset_id: str,
    cloud_document_ids: List[str],
    ndi_dataset: Optional['Dataset'] = None,
    sync_options: Optional['SyncOptions'] = None
) -> List['Document']:
    """Download a collection of NDI documents and their files.

    This function downloads document metadata from the cloud, and if
    sync_options.sync_files is True, it also downloads the associated data files
    to a local staging location and updates document file information.
    Finally, it adds the documents to the local NDI dataset if provided.

    Args:
        cloud_dataset_id: The ID of the dataset on the cloud
        cloud_document_ids: A list of cloud-specific document IDs to download
        ndi_dataset: The local NDI dataset object (optional)
        sync_options: Synchronization options (optional)

    Returns:
        A list of the ndi.document objects that were downloaded and added
        to the dataset
    """
    from ndi.cloud.download.download_document_collection import download_document_collection
    from ndi.cloud.download.download_dataset_files import download_dataset_files
    from ndi.docs import docfun

    if sync_options is None:
        from ndi.cloud.sync.sync_options import SyncOptions
        sync_options = SyncOptions()

    downloaded_ndi_documents = []

    if not cloud_document_ids:
        if sync_options.verbose:
            print('No document IDs provided to download.')
        return downloaded_ndi_documents

    if sync_options.verbose:
        if not cloud_document_ids or cloud_document_ids == [""]:
            print('Attempting to download all documents...')
        else:
            print(f'Attempting to download {len(cloud_document_ids)} documents...')

    # 1. Download documents
    # This function should return a list of ndi.document objects
    new_ndi_documents = download_document_collection(cloud_dataset_id, cloud_document_ids)

    if not new_ndi_documents:
        print('Warning: No documents were retrieved from the cloud for the given IDs.')
        return downloaded_ndi_documents

    if sync_options.verbose:
        print(f'Successfully retrieved metadata for {len(new_ndi_documents)} documents.')

    # 2. Handle associated data files
    if sync_options.sync_files:
        if sync_options.verbose:
            print('SyncFiles is true. Processing associated data files...')

        if ndi_dataset is None:
            import tempfile
            root_files_folder = tempfile.gettempdir()
        else:
            root_files_folder = ndi_dataset.path

        # Todo: Ensure proper cleanup if anything goes wrong before files
        # are ingested to database.
        files_target_folder = os.path.join(root_files_folder, Constants.FILE_SYNC_LOCATION)

        file_uids_to_download = get_file_uids_from_documents(new_ndi_documents)

        if file_uids_to_download:
            if sync_options.verbose:
                print(f'Found {len(file_uids_to_download)} unique file UIDs to '
                      f'download for these documents.')
                print(f'Ensuring download directory exists: {files_target_folder}')

            if not os.path.isdir(files_target_folder):
                os.makedirs(files_target_folder, exist_ok=True)

            # This function should download files to the files_target_folder
            download_dataset_files(
                cloud_dataset_id,
                files_target_folder,
                file_uids_to_download,
                verbose=sync_options.verbose
            )

            if sync_options.verbose:
                print('Completed downloading data files.')

            # Update document file info to point to local files
            if sync_options.verbose:
                print('Updating document file info to point to local files.')
        else:
            if sync_options.verbose:
                print('No associated files found for these documents, or files '
                      'already local.')
                print('Updating document file info (SyncMode.Local, but no new '
                      'files to point to).')

        document_update_fcn = lambda doc: update_file_info_for_local_files(
            doc, files_target_folder
        )
    else:
        if sync_options.verbose:
            print('"SyncFiles" option is false. Updating document file info to '
                  'reflect remote files.')

        document_update_fcn = lambda doc: update_file_info_for_remote_files(
            doc, cloud_dataset_id
        )

    # 3. Update file info for documents based on local / remote location
    new_ndi_documents = docfun(document_update_fcn, new_ndi_documents)

    # 4. Add documents to the local dataset
    if ndi_dataset is not None:
        if sync_options.verbose:
            print(f'Adding {len(new_ndi_documents)} processed documents to the '
                  f'local dataset...')
        ndi_dataset.database_add(new_ndi_documents)
        if sync_options.verbose:
            print('Documents added to the dataset.')

    return new_ndi_documents
